### PR TITLE
Adds tooltips to /tg/c keywords. Adds support for chat embedded tgui components

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

[Refactors tcg code a bit. Adds support for "keywords"](https://github.com/tgstation/tgstation/commit/276f7269c27c5100e90523b884cf11ff8b22b961)

Idea is to allow card makers to embed tooltips in their card
descriptions.
These tooltips are defined in the keywords.json file
They can be referenced using {$keyword_name}

I've also done some refactoring to move more logic and state onto the
subsystem, and implemented a few keywords from the wiki

[Adds support for embedding react components in tgui chat](https://github.com/tgstation/tgstation/commit/80810d5c3b8f3f5e950a047821c9828ca5e3e39a)

This is done by adding the data-component attribute to an html element
The value of that attibute is the component you want to use.
New components can be added by modifying the TGUI_CHAT_COMPONENTS list
in tgui-panel/chat/renderer.js.

Props can also be passed in in a limited capacity.
Any props you wish to pass must be added to
TGUI_CHAT_ATTRIBUTES_TO_PROPS.
This is due to a style restriction of html attributes, they cannot
contain an upper case char.
Use this list to convert between attibute compatible text and the prop's
name.

Props support 3 datatypes.
true and false can be passed by wrapping them in ""s. (Note to self add
a special char here to prevent colison with people just passing the
string true.
Numbers are supported in a limited capacity. Whitespace is not allowed,
but floats and ints are fair game.
And of course, strings are fully supported.

I've currently added support for Tooltip, since that's what I'm using
this for. Also added some tooltip html styles to the chat css.

## Why It's Good For The Game

image
Discord_TJENJzCiIq

## Changelog

:cl:
qol: /tg/c cards now support in chat tooltips, finally you don't need to look up what Hivemind means for the 200th time
code: tgui chat now supports embedding components into it via html attributes. Go make the game look better lads.
/:cl:

### CODE REVIEW

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -113,6 +113,7 @@ Things you **CAN'T** do:
 
 [MC Tab Guide](./guides/MC_tab.md)
 
+[Embedding tGUI components in chat](../tgui/docs/chat-embedded-components.md)
```
Comment (Fikou) [inline .github/CONTRIBUTING.md:116]:
```suggestion
[Embedding tgui components in chat](../tgui/docs/chat-embedded-components.md)
```
its called tgui everywhere

**Comment by cacogen** [comment]
Could this be used to have features like Discord, such as being able to @ people?

**Review by stylemistake — COMMENTED** [review]
Nicely done!

File: code/__DEFINES/span.dm [hunk 2]
```diff
@@ -114,3 +114,7 @@
 #define span_warning(str) ("<span class='warning'>" + str + "</span>")
 #define span_yell(str) ("<span class='yell'>" + str + "</span>")
 #define span_yellowteamradio(str) ("<span class='yellowteamradio'>" + str + "</span>")
+
+// Spans that use embedded tgui components:
+// Sorted alphabetically
+#define span_tooltip(tip, main_text) ("<span data-component=\"Tooltip\" data-content=\"" + tip + "\"><span class=\"tooltip\">" + main_text + "</span></span>")
```
[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/__DEFINES/span.dm:120]: also, what happens when tip has " in it?

**Comment by stylemistake** [comment]
Although I should note that $true, $false for boolean is a very artificial syntax and I don't see why just "true" "false" can't be used.

File: code/__DEFINES/span.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/__DEFINES/span.dm:120]: I seem to recall requiring it, though I might be wrong. Will try

[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/__DEFINES/span.dm:120]: > also, what happens when tip has " in it?

I think you would need to double escape it, not sure if that would work tho

File: tgui/packages/tgui-panel/chat/renderer.js [hunk 3]
```diff
@@ -8,15 +8,29 @@ import { EventEmitter } from 'common/events';
...
 
 const logger = createLogger('chatRenderer');
 
 // We consider this as the smallest possible scroll offset
 // that is still trackable.
 const SCROLL_TRACKING_TOLERANCE = 24;
 
+// List of injectable component names to the actual type
+export const TGUI_CHAT_COMPONENTS = {
+  "Tooltip": Tooltip,
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui-panel/chat/renderer.js:24]: Would this work to turn strings into components?

**Comment by LemonInTheDark** [comment]
> Although I should note that $true, $false for boolean is a very artificial syntax and I don't see why just "true" "false" can't be used.

I didn't want to end up in a situation where someone wants to pass in `"true"`, but can't because it just keeps getting parsed to a bool
Is that handled by js's typing or something?

File: tgui/packages/tgui-panel/chat/renderer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (stylemistake) [reply tgui/packages/tgui-panel/chat/renderer.js:24]: Yes, this is functionally the same.

File: code/__DEFINES/span.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/__DEFINES/span.dm]: Doesn't, you were correct

**Review by stylemistake — APPROVED** [review]

**Pull request merged**